### PR TITLE
Hybrid store

### PR DESCRIPTION
**Pull request by jbrinley**
@thenbrent, I wanted your thoughts on this approach before I spend to much time going this route.

The Hybrid_Store would be used up until the migration is complete, and then would be replaced with the simpler DB_Store. It overrides those methods that would be searching for actions. First it delegates to the old (wp post) store to find anything that would match the query there. Those matching actions are immediately migrated to the new (custom table) store.

Theoretically, this has moved all of the actions that should match the query into the new store, so we can at this point delegate the request and get a canonical result.

Possible concerns:

1. Does this seem logically sound? Are there cases I'm missing where we might need results from the old store?

2. What are the performance implications of doing small migration batches at run time? It would obviously slow things down a bit, but this is in lieu of shutting down action processing entirely.

Any thoughts on the above, or other concerns?

### CODE REVIEW

**Comment by thenbrent** [comment]
I'm going to do a lot of thinking out loud below to try cover all cases on this.

The tl;dr is: I don't think anything needs to change and this is good to merge as is.

Please read over my thoughts below and let me know if you do think any of them do need to be addressed.

---

> Does this seem logically sound?

It does yes. 💯 

This is really amazing work. I wasn't actually sure if the general concept of having data in the old store and the new could work in a sane way until seeing this code. Seeing the `Hybrid_Store`, I now think not only that it will work, but that it is the best approach.

> What are the performance implications of doing small migration batches at run time?

There's a few cases where Action Scheduler will run these functions and the corresponding migration:

1. processing a queue, which happens in the background
2. viewing the Admin UI, which only impacts page load for site admins
3. when an event occurs which requires an action to be unscheduled or checked - this could occur anywhere and any time, including front-end facing requests

The last of those is my only concern. However, it's a very minor concern.

The write queries to the new DB will be very fast. The major perf issues in Action Scheduler will continue to be the read/write queries through WordPress APIs for posts, which need to happen regardless of the migration happening or not. I suspect in all of those front-end facing requests, the performance hit in writing to the custom tables on large sites would be insignificant compared to the overhead of the custom post type operations.

> Are there cases I'm missing where we might need results from the old store?

I see all methods from `fetch_action()` onwards down assume the data will be in the new store.

This is safe internally because none of those methods are called by Action Scheduler without first going through a search method first to find the `$action_id` (or `$claim`).

But I see a couple of potential issues with that. Both minor, so probably not worth addressing, but worth mentioning for posterity here I think.

1. Firstly, this tends to mix knowledge from `ActionScheduler_QueueRunner` and other classes in the `Hybrid_Store`, breaking the neat encapsulation of storage. Specifically, those methods in `Hybrid_Store` now assume they will only be called with an `$action_id` (or `$claim`) derived from `Hybrid_Store::find_action()` or `Hybrid_Store::query_actions()` (or `Hybrid_Store::stake_claim()`, or after one of those methods has been called for that action, which isn't really knowledge of the store. Theoretically, the `$action_id` _could_ come from another source.
2. Secondly, those methods are all `public`, so while they're not considered supported APIs for working with actions, it is entirely possible some badly behaving code is calling them directly, e.g. `ActionScheduler_Store::instance()->cancel_action( $action_id )`. It's likely safe in all cases, but if we want to be 100% confident there will be no issues prior to the migration completing, those methods probably shouldn't assume the `$action_id` or `$claim` will definitely be in the new store already. (There is a real example of badly behaving code calling one of these methods directly with [`ActionScheduler_Store::instance()-> get_claim_count()`](https://github.com/Prospress/action-scheduler/blob/1.5.1/classes/ActionScheduler_AdminView.php#L318). That will be fixed up, and it would work correctly even without being fixed up given the queues will all be in the new store, but it's worth mentioning as a real example already).

The difficulty (or impossibility?) in addressing these issues comes from ID conflicts. For example, if there are 10,000 actions in the _old/secondary_ store, and one of those action has a post ID of `1235`, then both `Hybrid_Store->secondary_store->fetch_action( 1235 )` and  `Hybrid_Store->primary_store->fetch_action( 1235 )` could return a valid action. However, those actions will be for different events, and the method can't know for sure which action the calling code was referring too.

For this reason, I think it's most sane to avoid changing them and always refer to the new/primary store.

> Any thoughts on the above, or other concerns?

My only other concern is the possibility of actions existing side-by-side for a long period time. This is beyond the scope of `Hybrid_Store`, and more in the realm of `Migration_Runner`, but worth mentioning briefly here.

If a store has actions spread out over years and migration only happens when fetching, claiming or querying actions, then it's possible those actions won't be migrated until they are actually run years later.

Ideally, a separate background migration process is also run to get actions migrated ASAP and avoid any unexpected issues with using the `Hybrid_Store` indefinitely. The `Migration_Command` takes care of this nicely for really large sites. But small sites will need this to happen automatically too.

**Comment by jbrinley** [comment]
Thanks for reviewing, @thenbrent. I'll move forward with the approach, based on your feedback, including tests and exploring for various edge cases. 

> The difficulty (or impossibility?) in addressing these issues comes from ID conflicts.

One possible approach to addressing this: before adding the first action to the new table, find the largest ID from the old store and start the auto-increment for the new store at the next int. We could store that dividing line in an option, and redirect methods to the appropriate queue based on a comparison of the ID with that marker. This breaks the encapsulation somewhat, but the situation would only come about when client code is already breaking said encapsulation.

> Ideally, a separate background migration process is also run to get actions migrated ASAP and avoid any unexpected issues with using the Hybrid_Store indefinitely.

Yes, that's a separate task I'm working on right now. I'll schedule a recurring job to run every five minutes (filterable) in batches of 100 (again, filterable) until everything is migrated to the new store. After the migration completes, the `DB_Store` becomes the default store, and `Hybrid_Store` will sit dormant.

**Comment by thenbrent** [comment]
> One possible approach to addressing this

I can't see any issues with that approach. 👍 

> I'll schedule a recurring job to run every five minutes (filterable) in batches of 100 (again, filterable) until everything is migrated to the new store. 

Perfect! ❤️
